extension/json/json_functions/json_transform.cpp

```diff
@@ -898,20 +898,32 @@ BoundCastInfo JSONToAnyCastBind(BindCastInput &input, const LogicalType &source,
 }
 
 void JSONFunctions::RegisterJSONTransformCastFunctions(CastFunctionSet &casts) {
-	auto json_to_any_cost = casts.ImplicitCastCost(JSONCommon::JSONType(), LogicalType::ANY);
-	casts.RegisterCastFunction(JSONCommon::JSONType(), LogicalType::ANY, JSONToAnyCastBind, json_to_any_cost);
-
-	const auto struct_type = LogicalType::STRUCT({{"any", LogicalType::ANY}});
-	auto json_to_struct_cost = casts.ImplicitCastCost(LogicalType::VARCHAR, struct_type) - 2;
-	casts.RegisterCastFunction(JSONCommon::JSONType(), struct_type, JSONToAnyCastBind, json_to_struct_cost);
-
-	const auto list_type = LogicalType::LIST(LogicalType::ANY);
-	auto json_to_list_cost = casts.ImplicitCastCost(LogicalType::VARCHAR, list_type) - 2;
-	casts.RegisterCastFunction(JSONCommon::JSONType(), list_type, JSONToAnyCastBind, json_to_list_cost);
-
-	const auto map_type = LogicalType::MAP(LogicalType::ANY, LogicalType::ANY);
-	auto json_to_map_cost = casts.ImplicitCastCost(LogicalType::VARCHAR, map_type) - 2;
-	casts.RegisterCastFunction(JSONCommon::JSONType(), map_type, JSONToAnyCastBind, json_to_map_cost);
+	// JSON can be cast to anything
+	for (const auto &type : LogicalType::AllTypes()) {
+		LogicalType target_type;
+		switch (type.id()) {
+		case LogicalTypeId::STRUCT:
+			target_type = LogicalType::STRUCT({{"any", LogicalType::ANY}});
+			break;
+		case LogicalTypeId::LIST:
+			target_type = LogicalType::LIST(LogicalType::ANY);
+			break;
+		case LogicalTypeId::MAP:
+			target_type = LogicalType::MAP(LogicalType::ANY, LogicalType::ANY);
+			break;
+		case LogicalTypeId::UNION:
+			target_type = LogicalType::UNION({{"any", LogicalType::ANY}});
+			break;
+		case LogicalTypeId::VARCHAR:
+			// We skip this one here as it's handled in json_functions.cpp
+			continue;
+		default:
+			target_type = type;
+		}
+		// Going from JSON to another type has the same cost as going from VARCHAR to that type
+		const auto json_to_target_cost = casts.ImplicitCastCost(JSONCommon::JSONType(), target_type);
```

Comment:
I'm not sure if this is actually the correct solution here. This will only allow JSON to be implicitly castable to built-in types (instead of to any types as before). Perhaps the correct solution is to increase the `json_to_any_cost`? Currently that is set to `1` which is very low and means we prefer casting JSON to anything **not** explicitly registered (i.e. we prefer casting JSON to decimal than to varchar, which seems like odd behavior to me).

The conflict in `MIN` comes from the fact that we can either cast `JSON` to `ANY`, or to `DECIMAL`, which both have the same low cost (cost: 1). 

In general I'm not entirely sure if we want JSON to be implicitly castable to all types in the first place. For example, is this behavior intentional?

```sql
select '42'::json + 84::bigint AS p
┌───────┐
│   p   │
│ int64 │
├───────┤
│   126 │
└───────┘
D select cos('42'::json) as c;
┌──────────────────────┐
│          c           │
│        double        │
├──────────────────────┤
│ -0.39998531498835127 │
└──────────────────────┘
```

Reply:
Thanks for the feedback. I agree that implicitly casting to any type is not desirable. I've made a minor change so that `JSON` is only implicitly castable to the same types as `VARCHAR` is implicitly castable to. This also makes it so that your example queries no longer work (because they don't work for `VARCHAR` either`).

`JSON` is still castable to `ANY`, because every type is castable to `ANY` by default. I've tried increasing this cost and still ran into ambiguity issues.

By registering all the cast functions like this we follow the same behavior as `VARCHAR`, which is more intuitive than what I had implemented before.